### PR TITLE
WI #1181, add Diagnostic error when a statement is across lines marke…

### PR DESCRIPTION
**Pull request by LANOYDO**
…d with debug and lines not marked debug

### CODE REVIEW

File: TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs [hunk 1]
```diff
@@ -289,6 +289,32 @@ public static void OnCodeElement(StartStatement statement, CodeElementsParser.St
...
+                int consumedTokensCount = codeElement.ConsumedTokens.Count;
+                if (consumedTokensCount > 1)
+                {
+                    bool isDebug = false, isNoDebug = false;
+                    for (int i = 0; i < consumedTokensCount; i++)
+                    {
+                        bool isDebugType = char.ToLower(codeElement.ConsumedTokens[i].TokensLine.IndicatorChar) == 'd';
+                        isDebug |= isDebugType;
+                        isNoDebug |= !isDebugType;
+                    }
```
Comment (mayanje) [inline TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs]: Why not stopping loop as soon as (isDebug && isNoDebug) is true? I think we can be an operator | is used.

File: TypeCobol.Test/Parser/Programs/Cobol85/DebugNormalLines.rdz.cbl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+       PROGRAM-ID. Pgm.
+       ENVIRONMENT DIVISION.
+       CONFIGURATION SECTION.
+       DATA DIVISION.
+       Working-Storage Section.
+       01 xxxx pic XXXX value 'toto'.
+       01 yyyy pic XXXX.
+       PROCEDURE DIVISION.
+           MOVE xxxx TO yyyy
+      d    DISPLAY 'name = '
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/Cobol85/DebugNormalLines.rdz.cbl]: done in Commits on Sep 30, 2019

File: TypeCobol/Compiler/Parser/CodeElementsParserStep.cs [hunk 3]
```diff
@@ -304,6 +304,21 @@ internal static IList<DocumentChange<ICodeElementsLine>> ParseProcessedTokensLin
                                     codeElement.ConsumedTokens = codeElement.ConsumedTokens.OrderBy(t => t.Line)
                                         .ThenBy(t => t.Column).ToList();
                                 }
+
+                                // #1181 : Detect CodeElement with a mix of Debug and "Normal" lines
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/Parser/CodeElementsParserStep.cs]: done in Commits on Sep 30, 2019

File: TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs (hunk 1 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/tree/95744efb6e12dc402f62f7c87de96a65cf496b6a

File: TypeCobol.Test/Parser/Programs/Cobol85/DebugNormalLines.rdzPGM.txt [hunk 4]
```diff
@@ -0,0 +1,35 @@
+--- Diagnostics ---
+Line 14[12,20] <27, Error, Syntax> - Syntax error : In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.
+Line 19[12,19] <27, Error, Syntax> - Syntax error : In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.
+Line 24[12,19] <27, Error, Syntax> - Syntax error : In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.
+Line 29[12,19] <27, Error, Syntax> - Syntax error : In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.
+Line 35[12,18] <27, Error, Syntax> - Syntax error : In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.
+Line 40[12,17] <27, Error, Syntax> - Syntax error : In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.
+
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/Cobol85/DebugNormalLines.rdzPGM.txt]: No error reported on line 58 in stacked program Pgm2 ?

File: TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs [hunk 5]
```diff
@@ -297,5 +297,32 @@ public static void OnCodeElement(AlterStatement statement, CodeElementsParser.Al
...
+                {
+                    bool isDebug = false, isNoDebug = false;
+                    for (int i = 0; i < consumedTokensCount; i++)
+                    {
+                        bool isDebugType = char.ToLower(codeElement.ConsumedTokens[i].TokensLine.IndicatorChar) == 'd';
+                        isDebug |= isDebugType;
+                        isNoDebug |= !isDebugType;
+                        if (isDebug && isNoDebug)
+                        {
+                            DiagnosticUtils.AddError(codeElement, "In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.");
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs]: I think the message "In debugging mode, a statement cannot span across lines marked with debug and lines not marked debug." sounds more natural.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 6]
```diff
@@ -657,6 +655,23 @@ private static void CheckMultipleFormComParam(CodeElement codeElement)
...
+            isCurrentNoDebug |= !isDebugType;
+            if (isCurrentDebug && isCurrentNoDebug)
+            {
+                DiagnosticUtils.AddError(codeElement,
+                    "In debugging mode, a statement cannot be across lines marked with debug and lines not marked debug.");
+                return true;
+            }
+
+            return false;
+        }
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: Please remove this redundant code, the method is not called.

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs [hunk 7]
```diff
@@ -18,7 +18,8 @@ namespace TypeCobol.Compiler.Parser
 	/// <summary>Builds a CodeElement object while visiting its parse tree.</summary>
 	internal partial class CodeElementBuilder: CodeElementsBaseListener {
 
-		private ParserRuleContext Context;
+	    private bool DebuggingMode { get; set; }
```
Comment (fm-117) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs]: Would be clearer if renamed to 'IsDebuggingModeEnabled'

File: TypeCobol.Test/Parser/Programs/Cobol85/DebugNormalLines.rdzPGM.txt (hunk 4 above)
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/Cobol85/DebugNormalLines.rdzPGM.txt]: Program Pgm2 is not in debugging mode

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs (hunk 6 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/tree/8adf8c367cfdcd4aaa32b91135b29bd72e31b660

File: TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs (hunk 5 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/CodeElementCheckers.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/tree/8adf8c367cfdcd4aaa32b91135b29bd72e31b660

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs (hunk 7 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/tree/8adf8c367cfdcd4aaa32b91135b29bd72e31b660

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs [hunk 8]
```diff
@@ -18,7 +18,8 @@ namespace TypeCobol.Compiler.Parser
 	/// <summary>Builds a CodeElement object while visiting its parse tree.</summary>
 	internal partial class CodeElementBuilder: CodeElementsBaseListener {
 
-		private ParserRuleContext Context;
+	    private bool IsDebuggingModeEnabled { get; set; }
+        private ParserRuleContext Context;
```
Comment (smedilol) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs]: Indentation is not correct.

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs [hunk 9]
```diff
@@ -280,6 +283,7 @@ public override void EnterSourceComputerParagraph(CodeElementsParser.SourceCompu
 			}
 			if(context.DEBUGGING() != null) {
 				paragraph.DebuggingMode = new SyntaxProperty<bool>(true, ParseTreeUtils.GetFirstToken(context.DEBUGGING()));
+                IsDebuggingModeEnabled = true;
```
Comment (smedilol) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs]: Indentation is not correct.

**Pull request merged**